pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java

```diff
@@ -474,7 +475,13 @@ public void registerListener(TopicName topicName, TopicPolicyListener<TopicPolic
 
     @Override
     public void unregisterListener(TopicName topicName, TopicPolicyListener<TopicPolicies> listener) {
-        listeners.computeIfAbsent(topicName, k -> Lists.newCopyOnWriteArrayList()).remove(listener);
+        List<TopicPolicyListener<TopicPolicies>> topicListeners = listeners.get(topicName);
+        if (topicListeners != null && topicListeners.remove(listener)) {
+            //delete topic from listeners if it's empty.
+            if (topicListeners.isEmpty()) {
```

Comment:
Now I understand that this code will work, but I find it quite convoluted.
so actually you can remove this "if" condition

Personally I would go with the `computeIfAbsent` way, but probably it is only a matter of taste.


A unit test around this class is really needed here :-)

Reply:
The difference is that 'isEmpty' in ConcurrentHashMap just count the map size, no lock required. But the `remove()` will find the key first and there is a lock as it's a modification ops.

By `computeIfAbsent`, do you mean like `map.computeIfAbsent(k,k->{... map.remove ...})`. Actually it's forbidden that call any modification method in the computeIfAbsent. There is a risk of dead lock issue. See here:

![image](https://user-images.githubusercontent.com/2770146/141292194-fe1639e8-428d-433e-a0d7-691f0f7de872.png)

Reply:
And I am working on unit test now. Writing unit test for concurrent problem is tricky.

Reply:
no need to write a concurrent unit test
just test basic behaviour
we do not have to test ConcurrentHashMap, it already works well...

Reply:
for computeIfAbsent...
if you return "null" it should simply not add anything to the map

Reply:
> for computeIfAbsent... if you return "null" it should simply not add anything to the map

Thx! Great to know. Got a little mixed with `org.apache.pulsar.common.util.collections.ConcurrentOpenHashMap`, in which a key with null value will be inserted if return "null" in computeIfAbsent.
See #12729